File: Peare/Resources/RT_MESSAGE/RT_MESSAGE.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading.Tasks;

namespace Peare
{
    public static class RT_MESSAGE
    {
        // --- Structures for Windows Message Table ---
        // These are essential and MUST accurately reflect the WinAPI definitions.
        // LayoutKind.Sequential is crucial for correct memory alignment.

        // MESSAGE_RESOURCE_HEADER is part of MESSAGE_RESOURCE_DATA
        [StructLayout(LayoutKind.Sequential)]
        public struct MESSAGE_RESOURCE_HEADER
        {
            public uint NumberOfBlocks; // Number of MESSAGE_RESOURCE_BLOCK structures
                                        // Followed by MESSAGE_RESOURCE_BLOCKs
        }

        // MESSAGE_RESOURCE_BLOCK defines a range of message IDs and their offset
        [StructLayout(LayoutKind.Sequential)]
        public struct MESSAGE_RESOURCE_BLOCK
        {
            public uint LowId;          // Lowest message ID in this block
            public uint HighId;         // Highest message ID in this block
            public uint OffsetToEntries; // Offset from the start of MESSAGE_RESOURCE_DATA to the first MESSAGE_RESOURCE_ENTRY for this block
        }

        // MESSAGE_RESOURCE_ENTRY is not explicitly a separate struct in the file,
        // but its components are always a short Length, short Flags, and then the text.
        // The 'length' field in MESSAGE_RESOURCE_ENTRY includes the 4-byte header (Length + Flags)
        // AND the string content AND its 2-byte null terminator.

        public static string Get(byte[] data)
        {
            ushort cp = 20127; // Default ASCII

            int offset = 0; // Current read position in the 'data' array
            if (Program.isOS2)
            {
                // First two bytes in OS/2 are the codepage
                cp = RT_STRING.ReadUInt16(data, ref offset);
            }

            StringBuilder output = new StringBuilder();
            output.AppendLine("MESSAGETABLE");
            output.AppendLine("{");

            // We need to manage a memory pointer for Marshal operations,
            // as data is a managed byte array.
            IntPtr unmanagedDataPtr = IntPtr.Zero;

            try
            {
                // Allocate unmanaged memory and copy the entire byte array to it.
                // This allows Marshal.PtrToStructure to work directly.
                unmanagedDataPtr = Marshal.AllocHGlobal(data.Length);
                Marshal.Copy(data, 0, unmanagedDataPtr, data.Length);

                // Current pointer for reading through the unmanaged memory block
                IntPtr currentUnmanagedPtr = unmanagedDataPtr;

                while (offset + 1 < data.Length) // This loop condition might need adjustment based on full resource parsing
                {
                    int msgId; // Will be the actual message ID from MESSAGE_RESOURCE_BLOCK/ENTRY
                    if (Program.currentHeaderType.StartsWith("PE"))
                    {
                        // --- Windows PE (RT_MESSAGETABLE) Parsing Logic ---

                        // The first thing in the MESSAGE_RESOURCE_DATA is MESSAGE_RESOURCE_HEADER.NumberOfBlocks
                        // Read NumberOfBlocks (assuming offset is at the beginning of the MESSAGE_RESOURCE_DATA)
                        uint numberOfBlocks = (uint)Marshal.ReadInt32(currentUnmanagedPtr, offset);
                        offset += Marshal.SizeOf<uint>(); // Advance offset past NumberOfBlocks

                        // The blocks array starts immediately after NumberOfBlocks
                        IntPtr blockPtr = IntPtr.Add(currentUnmanagedPtr, offset);
                        int blockSize = Marshal.SizeOf<MESSAGE_RESOURCE_BLOCK>();

                        for (int i = 0; i < numberOfBlocks; i++)
                        {
                            // Read the current MESSAGE_RESOURCE_BLOCK
                            MESSAGE_RESOURCE_BLOCK block = Marshal.PtrToStructure<MESSAGE_RESOURCE_BLOCK>(blockPtr);

                            // The OffsetToEntries is relative to the start of the MESSAGE_RESOURCE_DATA
                            // So, entryPtr points to the beginning of the MESSAGE_RESOURCE_ENTRY for this block
                            IntPtr entryPtr = IntPtr.Add(unmanagedDataPtr, (int)block.OffsetToEntries);

                            for (uint id = block.LowId; id <= block.HighId; id++)
                            {
                                // Read MESSAGE_RESOURCE_ENTRY fields
                                short entryLength = Marshal.ReadInt16(entryPtr); // Total length of this entry (including its header and string with null terminator)
                                short flags = Marshal.ReadInt16(entryPtr, 2);   // Flags (0 = ANSI, 1 = Unicode)

                                // The actual string data starts 4 bytes after the beginning of the entry
                                IntPtr textPtr = IntPtr.Add(entryPtr, 4);

                                string message = "ERROR: Could not decode message."; // Default error message

                                // Determine encoding based on flags
                                if (flags == 0) // ANSI (single-byte characters, double-byte null terminator)
                                {
                                    // Marshal.PtrToStringAnsi will read until the first 0x00.
                                    // The string content is single-byte, followed by 0x00 0x00.
                                    // PtrToStringAnsi correctly handles the 0x00 as a terminator.
                                    message = Marshal.PtrToStringAnsi(textPtr);
                                }
                                else if (flags == 1) // Unicode (UTF-16LE characters, double-byte null terminator)
                                {
                                    // Marshal.PtrToStringUni will read until the first 0x00 0x00.
                                    // This is the standard for Unicode strings in Windows resources.
                                    message = Marshal.PtrToStringUni(textPtr);
                                }
                                else
                                {
                                    // Handle unknown flags if necessary, this might indicate corrupted data
                                    message = $"UNKNOWN_FLAGS_{flags}_FOR_ID_{id}";
                                }

                                // Clean up the message text: remove carriage returns and line feeds
                                message = message.Replace("\r\n", "");

                                output.AppendLine($"\t0x{id:X4}, \"{message}\"");

                                // Advance to the next MESSAGE_RESOURCE_ENTRY using its total length
                                entryPtr = IntPtr.Add(entryPtr, entryLength);
                            }
                            // Advance to the next MESSAGE_RESOURCE_BLOCK
                            blockPtr = IntPtr.Add(blockPtr, blockSize);
                        }
                        // Once all blocks are processed, we should have read the entire Message Table.
                        // Break out of the outer while loop as we've finished processing this PE resource.
                        offset = data.Length; // Force exit from outer while loop
                        break; // Exit the loop for PE processing
                    }
                    else // Existing non-PE logic
                    {
                        msgId = RT_STRING.ReadByte(data, ref offset);
                        string message = RT_STRING.ReadNullTerminatedString(data, ref offset, cp);
                        output.AppendLine($"\t0x{msgId:X4}, \"{message}\"");
                    }
                }
            }
            finally
            {
                // Crucially, free the unmanaged memory when done.
                if (unmanagedDataPtr != IntPtr.Zero)
                {
                    Marshal.FreeHGlobal(unmanagedDataPtr);
                }
            }

            output.AppendLine("}");
            return output.ToString();
        }
    }
}

File: src/decoder/resources/RT_MESSAGE/RT_MESSAGE.cpp
#include "RT_MESSAGE.h"

#include "../RT_STRING/RT_STRING.h"

#include <QStringList>
#include <QTextCodec>
#include <stdexcept>

namespace peare {
namespace resources {

// --- Structures for Windows Message Table ---
// These are essential and MUST accurately reflect the WinAPI definitions.
// In Qt we read the fields explicitly in little-endian order instead of relying on
// LayoutKind.Sequential and Marshal.PtrToStructure.

// MESSAGE_RESOURCE_HEADER is part of MESSAGE_RESOURCE_DATA.
// It contains NumberOfBlocks and is followed by MESSAGE_RESOURCE_BLOCKs.

// MESSAGE_RESOURCE_BLOCK defines a range of message IDs and their offset:
// LowId, HighId and OffsetToEntries.

// MESSAGE_RESOURCE_ENTRY is not explicitly a separate struct in the file,
// but its components are always a short Length, short Flags, and then the text.
// The Length field includes the 4-byte header (Length + Flags), the string content
// and its null terminator.
namespace {

quint16 readLe16(const QByteArray& data, qsizetype offset)
{
    if (offset < 0 || offset + 2 > data.size())
        throw std::runtime_error("Attempted to read past end of message data.");

    const auto* bytes = reinterpret_cast<const uchar*>(data.constData() + offset);
    return quint16(bytes[0]) | (quint16(bytes[1]) << 8);
}

quint32 readLe32(const QByteArray& data, qsizetype offset)
{
    if (offset < 0 || offset + 4 > data.size())
        throw std::runtime_error("Attempted to read past end of message data.");

    const auto* bytes = reinterpret_cast<const uchar*>(data.constData() + offset);
    return quint32(bytes[0])
        | (quint32(bytes[1]) << 8)
        | (quint32(bytes[2]) << 16)
        | (quint32(bytes[3]) << 24);
}

QString cleanMessage(QString message)
{
    // Clean up the message text: remove carriage returns and line feeds.
    message.replace(QStringLiteral("\r\n"), QString());
    return message;
}

QString decodeAnsiEntry(const QByteArray& data, qsizetype offset, qsizetype byteCount)
{
    if (byteCount < 0 || offset < 0 || offset + byteCount > data.size())
        throw std::runtime_error("Invalid ANSI message entry length.");

    QByteArray bytes = data.mid(offset, byteCount);
    const qsizetype nullIndex = bytes.indexOf('\0');
    if (nullIndex >= 0)
        bytes.truncate(nullIndex);
    return QString::fromLocal8Bit(bytes);
}

QString decodeUnicodeEntry(const QByteArray& data, qsizetype offset, qsizetype byteCount)
{
    if (byteCount < 0 || offset < 0 || offset + byteCount > data.size())
        throw std::runtime_error("Invalid Unicode message entry length.");

    byteCount -= byteCount % 2;
    QString value;
    value.reserve(int(byteCount / 2));

    for (qsizetype i = 0; i + 1 < byteCount; i += 2) {
        const quint16 ch = readLe16(data, offset + i);
        if (ch == 0)
            break;
        value.append(QChar(ch));
    }
    return value;
}

} // namespace

QString RT_MESSAGE::Get(const QByteArray& data,
                        ModuleFormat headerType,
                        bool isOs2)
{
    quint16 codePage = 20127; // Default ASCII
    qsizetype offset = 0; // Current read position in the data array

    if (((headerType == ModuleFormat::LE || headerType == ModuleFormat::NE) && isOs2)
        || headerType == ModuleFormat::LX)
    {
        // OS/2 1.x resources may omit the codepage prefix.
        codePage = 850;
        if (data.size() >= 2) {
            const quint16 candidate = quint16(uchar(data.at(0))) | (quint16(uchar(data.at(1))) << 8);
            const QByteArray codecName = QByteArray("IBM ") + QByteArray::number(candidate);
            if (candidate != 0 && QTextCodec::codecForName(codecName) != nullptr) {
                codePage = candidate;
                offset = 2;
            }
        }
    }

    QStringList output;
    output << QStringLiteral("MESSAGETABLE") << QStringLiteral("{");

    if (headerType == ModuleFormat::PE) {
        // --- Windows PE (RT_MESSAGETABLE) Parsing Logic ---
        if (data.size() < 4)
            throw std::runtime_error("Invalid MESSAGE_RESOURCE_DATA header.");

        // The first thing in MESSAGE_RESOURCE_DATA is MESSAGE_RESOURCE_HEADER.NumberOfBlocks.
        const quint32 numberOfBlocks = readLe32(data, 0);
        const quint64 blockTableEnd = 4ULL + quint64(numberOfBlocks) * 12ULL;
        if (blockTableEnd > quint64(data.size()))
            throw std::runtime_error("Incomplete MESSAGE_RESOURCE_BLOCK table.");

        // The blocks array starts immediately after NumberOfBlocks.
        qsizetype blockOffset = 4;
        for (quint32 blockIndex = 0; blockIndex < numberOfBlocks; ++blockIndex) {
            const quint32 lowId = readLe32(data, blockOffset);
            const quint32 highId = readLe32(data, blockOffset + 4);
            const quint32 entriesOffset = readLe32(data, blockOffset + 8);
            blockOffset += 12;

            if (highId < lowId)
                throw std::runtime_error("Invalid message ID range.");

            // OffsetToEntries is relative to the start of MESSAGE_RESOURCE_DATA.
            qsizetype entryOffset = qsizetype(entriesOffset);
            for (quint32 id = lowId; id <= highId; ++id) {
                if (entryOffset < 0 || entryOffset + 4 > data.size())
                    throw std::runtime_error("Incomplete MESSAGE_RESOURCE_ENTRY header.");

                // Read MESSAGE_RESOURCE_ENTRY fields.
                const quint16 entryLength = readLe16(data, entryOffset);
                const quint16 flags = readLe16(data, entryOffset + 2);
                if (entryLength < 4 || entryOffset + entryLength > data.size())
                    throw std::runtime_error("Invalid MESSAGE_RESOURCE_ENTRY length.");

                // The actual string data starts 4 bytes after the beginning of the entry.
                const qsizetype textOffset = entryOffset + 4;
                const qsizetype textLength = entryLength - 4;
                QString message;

                // Determine encoding based on flags.
                if (flags == 0) { // ANSI
                    message = decodeAnsiEntry(data, textOffset, textLength);
                } else if (flags == 1) { // Unicode (UTF-16LE)
                    message = decodeUnicodeEntry(data, textOffset, textLength);
                } else {
                    message = QStringLiteral("UNKNOWN_FLAGS_%1_FOR_ID_%2")
                                  .arg(flags)
                                  .arg(id);
                }

                output << QStringLiteral("\t0x%1, \"%2\"")
                              .arg(id, 4, 16, QLatin1Char('0'))
                              .arg(cleanMessage(message));
                // Advance to the next MESSAGE_RESOURCE_ENTRY using its total length.
                entryOffset += entryLength;

                if (id == 0xFFFFFFFFu)
                    break;
            }
        }
    } else { // Existing non-PE logic
        while (offset + 1 < data.size()) {
            const quint8 messageId = RT_STRING::ReadByte(data, offset);
            const QString message = RT_STRING::ReadNullTerminatedString(data, offset, codePage);
            output << QStringLiteral("\t0x%1, \"%2\"")
                          .arg(messageId, 4, 16, QLatin1Char('0'))
                          .arg(cleanMessage(message));
        }
    }

    output << QStringLiteral("}");
    return output.join(QLatin1Char('\n')) + QLatin1Char('\n');
}

ResourcePreview RT_MESSAGE::preview(const ResourceEntry& entry)
{
    ResourcePreview preview;
    try {
        preview.text = Get(entry.data, entry.format, entry.isOs2);
    } catch (const std::exception& error) {
        preview.error = QString::fromUtf8(error.what());
    }
    return preview;
}

} // namespace resources
} // namespace peare
